### PR TITLE
Brave 4.x modernization

### LINKED ISSUE #264: Make BraveServletFilter customizable
I started including brave in our projects and works great. However, we faced some issues when trying to customize the behavior a little bit and ended up copy pasting the complete implementation.

BraveServletFilter.doFilter is to large, the creation of ServletHttpServerRequest and HttpServerRequestAdapter should be moved into a dedicated method that would allow subclassing and modifying it. Or maybe even better would be to be able to set a factory. We modified HttpServerResponseAdapter to provide more than just the TraceKeys.HTTP_STATUS_CODE.

I can do a PR for this.

### PR DESCRIPTION
**Pull request by adriancole**
This module a redo of all major instrumentation libraries since Brave 3.
Artifacts have the naming convention "brave-instrumentation-XXX": for
example, the directory "servlet" includes the artifact "brave-instrumentation-servlet".

Notably, this adds a more `HttpTracing` component to configure http client
and server tracing. It also adds support for async servlet and apache http
client, and tests edge cases like multiple servlet versions. Finally, it
tests and benchmarks every http component, so that people are aware of
overhead is involved in tracing.

Here's an example of configuring OkHttp. Note that all instrumentation
have the same naming policy TracingXXX, where XXX is usually the same
as the type returned.

```java
tracing = Tracing.newBuilder()
                 .localServiceName("my-service")
                 .reporter(reporter)
                 .build();
httpTracing = HttpTracing.newBuilder(tracing).serverName("github").build();
okhttp = TracingCallFactory.create(httpTracing, new OkHttpClient());
```

### Http tracing
Most instrumentation are based on http communication. For this reason,
we have specialized handlers for http clients and servers. All of these
are configured with `HttpTracing`.

The `HttpTracing` class holds a reference to a tracing component and also
includes instructions on what to put into http spans.

By default, the following is added for both http clients and servers:
* Span.name as the http method in lowercase: ex "get"
* Tags/binary annotations:
  * "http.path", which does not include query parameters.
  * "http.status_code" when the status us not success.
* Remote IP and port information

Naming and tags are configurable in a library-agnostic way. For example,
the same `HttpTracing` component configures OkHttp or Apache HttpClient
identically.

For example, to change the span and tag naming policy for clients, you
can do something like this:

```java
httpTracing = httpTracing.toBuilder()
    .clientParser(new HttpClientParser() {
      @Override public <Req> String spanName(HttpAdapter<Req, ?> adapter, Req req) {
        return adapter.method(req).toLowerCase() + " " + adapter.path(req);
      }

      @Override
      public <Req> void requestTags(HttpAdapter<Req, ?> adapter, Req req, brave.Span span) {
        span.tag(TraceKeys.HTTP_URL, adapter.url(req)); // the whole url, not just the path
      }
    })
    .serverName("remote-service") // assume both libraries are calling the same service
    .build();

apache = TracingHttpClientBuilder.create(httpTracing).build();
okhttp = TracingCallFactory.create(httpTracing, new OkHttpClient());
```

---
Work accomplished in support of this
 - [x] Rough in a tentative client and server tracer api
 - [x] Integrate each library such that it sets spans in scope so `Tracer.currentSpan()` works
 - [x] Integrate each staticly initialized library such that it uses `Tracer.current()`
 - [x] Backport client and server integration tests
 - [x] add Async HttpClient and Servlet (which we never had before)
 - [x] add Tests for servlet 2.5 and grpc compat (which we never had before)
 - [x] ensure we can swap out propagation format when needed. Ex https://github.com/TraceContext/tracecontext-spec
 - [x] revise tagging approach so that it solves #276 #264
 - [x] Add tests for http client redirects
 - [x] Revisit exception parsing
 - [x] Revisit remote address parsing

Fixes #276
Fixes #264

### CODE REVIEW

**Comment by adriancole** [comment]
waiting a day for feedback on https://github.com/openzipkin/brave/issues/166#issuecomment-277634798 before proceeding to implement with the propagation design in mind

File: brave-apache-http-interceptors/src/main/java/brave/httpclient4/HttpClient4Instrumentation.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import org.apache.http.HttpResponseInterceptor;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import zipkin.TraceKeys;
+
+import static com.github.kristofa.brave.internal.Util.checkNotNull;
+
+public final class HttpClient4Instrumentation {
```
Comment (adriancole) [inline brave-apache-http-interceptors/src/main/java/brave/httpclient4/HttpClient4Instrumentation.java]: we've been doing this wrong :) if we want folks to use tracer.currentSpan we need to implement HttpProcessor so that we can scope the span before any of the interceptors fire.

that way, people can write an interceptor that uses the current span without needing to know a magic attribute. This is particularly important for slf4j or log4j2 mdc logging. If we set the span ID before any interceptors go, anything that happens will end up with the same ids in the log. The only way I can see how to do this is implementing a TracingHttpProcessor

cc @shakuzen

Comment (adriancole) [inline brave-apache-http-interceptors/src/main/java/brave/httpclient4/HttpClient4Instrumentation.java]: actually this won't work because there's a lot of logic inside the HttpClientBuilder which is only invoked if HttpProcessor isn't overridden. IOTW, I don't think we'll be able to neatly do implicit propagation. Rather, we'd need to document the attribute we use internally if someone wants to access the current span.

File: brave-grpc/src/test/java/brave/grpc/ITBraveClientInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+import zipkin.Constants;
+import zipkin.Endpoint;
+import zipkin.Span;
+import zipkin.storage.InMemoryStorage;
+
+import static java.util.Arrays.asList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.failBecauseExceptionWasNotThrown;
+
+public class ITBraveClientInterceptor {
```
Comment (adriancole) [inline brave-grpc/src/test/java/brave/grpc/ITBraveClientInterceptor.java]: Added this which is a port of the http test

**Comment by adriancole** [comment]
added jaxrs2 client impl and tests

**Comment by adriancole** [comment]
ok all client instrumentation has been retouched except the database ones.. I'm kindof going to save "best for last" on that one. Moving on to server instrumentation

note: pay little attention to the "Config" types defined, they haven't been thought through and are subject to change. This is mainly getting all tests passing and exercising the "current span" apis

**Comment by adriancole** [comment]
added some of the server instrumentation cc @kmaci3k

note: this PR needs one change to the "current span" PR it depends on #369, which is a replacement for BraveExecutorService (a couple http client libraries need it and tests are disabled until that's in)

**Comment by adriancole** [comment]
all instrumentation except jdbc, mysql and cxf have been redone. I will probably do cxf just to prove the api approach works

File: brave-http-tests/src/main/java/brave/http/StrictCurrentTraceContext.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package brave.http;
+
+import brave.propagation.CurrentTraceContext;
+import brave.propagation.TraceContext;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class StrictCurrentTraceContext implements CurrentTraceContext {
```
Comment (adriancole) [inline brave-http-tests/src/main/java/brave/http/StrictCurrentTraceContext.java]: Added this as the default implementation being an inheritable thread local makes it hard to test if propagation works by accident (ex only because inheritable thread local is used), or if because it was coded correctly. You'll notice this also is defensive about being closed on the same thread as it was opened.

**Comment by adriancole** [comment]
rebased on top of #369

**Comment by adriancole** [comment]
NEXT STEPS:
* implement sparkjava and database instrumentation with brave 4 apis
* take a closer look at instrumentation approach and generalize anything copy/pasted >3 times
* revise tagging approach so that it solves #276 #264

**Comment by adriancole** [comment]
UPDATE: implemented sparkjava and database instrumentation with brave 4 apis

noticed we should revisit #210 as the database drivers need a reference to a tracer post-initialization (static reference)

**Comment by adriancole** [comment]
ok rebased over some build things that were tripping up yesterday. will resume in a bit

**Comment by adriancole** [comment]
ok current code solves the JDBC problem via #379 

Moving on to other todos

**Comment by adriancole** [comment]
https://github.com/openzipkin/brave/pull/381 is needed so that we can wire in propagation handlers

**Comment by adriancole** [comment]
rebased over #381 but not refactored to take advantage of it (beyond making propagation type not explicit)

File: brave-apache-http-interceptors/src/main/java/brave/httpclient4/HttpClientTracing.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    return new HttpClientBuilder() {
+
+      /**
+       * protocol exec is the second in the execution chain, so is invoked before a request is
+       * provisioned. We provision and scope a span here, so that application interceptors can see
+       * it via {@link Tracer#currentSpan()}.
+       */
+      @Override protected ClientExecChain decorateProtocolExec(ClientExecChain exec) {
+        return (route, request, context, execAware) -> {
+          Span next = tracer.nextSpan();
```
Comment (pavolloffay) [inline brave-apache-http-interceptors/src/main/java/brave/httpclient4/HttpClientTracing.java]: It is a nice approach :+1: But It seems that this is run in a different thread.

**Comment by adriancole** [comment]
Ahah but if it did the test would fail!

This is not the async client, so it is blocking (same thread). Theres
another safety which is the strict context impl. Havent done async in this
library yet

On 22 Apr 2017 23:16, "Pavol Loffay" <notifications@github.com> wrote:

*@pavolloffay* commented on this pull request.
------------------------------

In brave-apache-http-interceptors/src/main/java/brave/httpclient4/
HttpClientTracing.java
<https://github.com/openzipkin/brave/pull/350#discussion_r112811778>:

> +  /**
+   * Preferred way for constructing http clients, as it allows use of
{@link Tracer#currentSpan}
+   *
+   * <p>Implicitly requires a minimum version of httpclient 4.3
+   */
+  public HttpClientBuilder httpClientBuilder() {
+    return new HttpClientBuilder() {
+
+      /**
+       * protocol exec is the second in the execution chain, so is
invoked before a request is
+       * provisioned. We provision and scope a span here, so that
application interceptors can see
+       * it via {@link Tracer#currentSpan()}.
+       */
+      @Override protected ClientExecChain
decorateProtocolExec(ClientExecChain exec) {
+        return (route, request, context, execAware) -> {
+          Span next = tracer.nextSpan();

It is a nice approach 👍 But It seems that this is run in a different
thread.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/openzipkin/brave/pull/350#pullrequestreview-34159439>,
or mute the thread

**Comment by pavolloffay** [comment]
@adriancole thanks, my fault I didn't use brave API properly. It's pity that this cannot be used for an async client.

**Comment by adriancole** [comment]
@adriancole <https://github.com/adriancole> thanks, my fault I didn't use
brave API properly. It's pity that this cannot be used for an async client


Async has different dep (different version number too annoyingly). Imho it
is better to align these with the primary jar.. so not a big deal to do
another type for async. Also that keeps complexity in the async type. Our
old interceptors didnt work well for async anyway.

File: brave-apache-http-interceptors/src/main/java/brave/httpclient4/HttpClientTracing.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        };
+      }
+
+      /**
+       * main exec is the first in the execution chain, so last to execute. This creates a concrete
+       * http request, so this is where the timing in the span occurs.
+       *
+       * <p>This ends the span (and scoping of it) created by {@link #decorateMainExec(ClientExecChain)}.
+       */
+      @Override protected ClientExecChain decorateMainExec(ClientExecChain exec) {
```
Comment (adriancole) [inline brave-apache-http-interceptors/src/main/java/brave/httpclient4/HttpClientTracing.java]: @llinder @devinsba @hyleung so when introducing support for custom parsing, I chose to keep generic parameters "raw" meaning they represent the actual native types. This is mostly to make the below code readable and without the clutter of `new FooRequestAdapter(request)` and `new FooResponseAdapter(response)`.

You'll also notice the adapter (maybe converter?) below is very straightforward.

File: brave-http-tests/src/main/java/brave/http/ITHttpClient.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+    assertThat(collectedSpans())
+        .extracting(s -> s.name)
+        .containsExactly("get");
+  }
+
+  @Test
+  public void supportsPortableCustomization() throws Exception {
+    close();
+    client = newClient(server.getPort(), new HttpClientParser<Req, Resp>(){
```
Comment (adriancole) [inline brave-http-tests/src/main/java/brave/http/ITHttpClient.java]: @llinder @devinsba @hyleung the tradeoffs come out here.. `HttpClientParser` is generically parameterized, but.. it can be used for any type. For example, the spanName below gives you the ability to adapt the raw type into a generic http model, but you can also access the raw request directly should you need to.

While the adapt (maybe we should call it convert or translate) is a bit explicit, it has the advantage of more concise and direct code in the instrumentation, and also a minor point which is that you don't need to allocate objects for every request and response (because this is converting an existing object vs wrapping it)

**Comment by adriancole** [comment]
started readying for final lap.. this included removing instrumentation that is very old (like jersey v1 as they can be asked for later).. I've moved all new instrumentation under the instrumentation folder and using brave-instrumentation-X which is similar naming to zipkin-js

here's what's left.
* rework client code to be like server code
* cleanup READMEs
* add javadoc
* get others to look at it.

File: brave-http-tests/src/main/java/brave/http/ITHttpClient.java (hunk 6 above)
Comment (llinder) [inline brave-http-tests/src/main/java/brave/http/ITHttpClient.java]: The pattern looks good to me and I can't think of anything I would change just looking at it.  Nice work so far!

**Comment by adriancole** [comment]
ok reworked the server code. other pending todos still todos

File: instrumentation/servlet/src/main/java/brave/servlet/HttpServletHandler.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import brave.http.HttpServerParser;
+import brave.internal.Nullable;
+import java.nio.ByteBuffer;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import zipkin.Endpoint;
+
+import static brave.internal.InetAddresses.ipStringToBytes;
+
+public final class HttpServletHandler // public for others like sparkjava to use
```
Comment (adriancole) [inline instrumentation/servlet/src/main/java/brave/servlet/HttpServletHandler.java]: @hyleung @llinder so all server impls will need to declare their own HttpAdapter, unless they share types with another library. server impls that have special knowledge of the client address may choose to make their own HttpServerHandler to accomodate that.

I hope looking at the servlet example here might help decide if extending similarly is worth it for ratpack. If you notice gaps, do tell regardless!

File: instrumentation/servlet/src/main/java/brave/servlet/HttpServletHandler.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    extends HttpServerHandler<HttpServletRequest, HttpServletResponse> {
+
+  public HttpServletHandler(HttpServerParser<HttpServletRequest, HttpServletResponse> parser) {
+    super(new HttpServletAdapter(), parser);
+  }
+
+  @Nullable public Endpoint.Builder remoteEndpoint(HttpServletRequest request) {
+    byte[] addressBytes = ipStringToBytes(request.getHeader("X-Forwarded-For"));
+    if (addressBytes == null) addressBytes = ipStringToBytes(request.getRemoteAddr());
```
Comment (adriancole) [inline instrumentation/servlet/src/main/java/brave/servlet/HttpServletHandler.java]: ex we are overriding here because servlet has a `getRemoteAddr` function which not all libraries have. The default we override looks at X-Forwarded-For only

File: instrumentation/jaxrs2/src/main/java/brave/jaxrs2/TracingContainerFilter.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.Provider;
+import zipkin.Constants;
+
+import static javax.ws.rs.RuntimeType.SERVER;
+
+@Provider
+@Priority(0) // to make the span in scope visible to other filters
+@ConstrainedTo(SERVER)
```
Comment (jplock) [inline instrumentation/jaxrs2/src/main/java/brave/jaxrs2/TracingContainerFilter.java]: should we add https://jersey.java.net/apidocs/2.12/jersey/javax/ws/rs/container/PreMatching.html on here as well so this provider happens before any routing decisions are made?

Comment (adriancole) [inline instrumentation/jaxrs2/src/main/java/brave/jaxrs2/TracingContainerFilter.java]: sure. will do and add a comment to the effect. thanks for your help!

Comment (adriancole) [inline instrumentation/jaxrs2/src/main/java/brave/jaxrs2/TracingContainerFilter.java]: hmm that breaks all the tests :P

**Comment by adriancole** [comment]
implemented async servlet and servlet 2.5 support. only thing left is async http client, look into jaxrs annotation things then cleanups.

**Comment by adriancole** [comment]
to test servlet 2.5 will be a minor pain (a maven invoker test, like we have in zipkin-server). I will definitely add this as otherwise it is punting to eyeballing and hoping it works.

File: instrumentation/servlet/src/main/java/brave/servlet/TracingFilter.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  final TraceContext.Extractor<HttpServletRequest> extractor;
+
+  TracingFilter(Builder builder) {
+    tracer = builder.tracing.tracer();
+    handler = new HttpServletHandler(builder.parser);
+    extractor = builder.tracing.propagation().extractor(HttpServletRequest::getHeader);
+  }
+
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain)
```
Comment (adriancole) [inline instrumentation/servlet/src/main/java/brave/servlet/TracingFilter.java]: @devinsba this servlet filter passes all tests sync and async. Will add an invoker test to make sure it also works on servlet 2.5 in practice (we have had people still using that, yes).

File: instrumentation/servlet/src/main/java/brave/servlet/TracingFilter.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  /** Creates a tracing filter with defaults. Use {@link #newBuilder(Tracing)} to customize. */
+  public static TracingFilter create(Tracing tracing) {
+    return newBuilder(tracing).build();
+  }
+
+  public static Builder newBuilder(Tracing tracing) {
+    return new Builder(tracing);
+  }
+
+  public static final class Builder {
```
Comment (adriancole) [inline instrumentation/servlet/src/main/java/brave/servlet/TracingFilter.java]: ps I looked at doing something like making a base builder class, but it made generics crazy.. it is more distracting to try and do that (or even autovalue) vs implement directly.

**Comment by adriancole** [comment]
ok servlet 2.5 tests are in (via maven invoker). Glad I added them as I had a bug in my class detection stuff.

**Comment by adriancole** [comment]
next thing I have is the apache async http client (mostly to show it can work).

I'm pretty happy with the api, but will likely polish it a little after stepping back and looking at each instrumentation closely.

File: instrumentation/servlet/src/main/java/brave/servlet/TracingFilter.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public final class TracingFilter implements Filter {
```
Comment (adriancole) [inline instrumentation/servlet/src/main/java/brave/servlet/TracingFilter.java]: @johnament FYI there will be some adjustments tomorrow (likely HttpTracing, which will be a component that includes Tracing and default http client and server tagging policy). However, most of this code will look the same. We won't be adding new Brave 3 style instrumentation moving forward. If you want feign or CDI, it would be based on the new stuff. Have a look!

**Comment by adriancole** [comment]
ok added HttpTracing component, which I think dramatically simplifies construction of instrumentation. Most are less than 100lines all-in

File: brave/src/main/java/brave/http/HttpTracing.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
+package brave.http;
+
+import brave.Tracing;
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class HttpTracing {
```
Comment (adriancole) [inline brave/src/main/java/brave/http/HttpTracing.java]: @llinder @jplock @hyleung @johnament @takezoe @devinsba so I think this is the latest thing to use in injection/configuraion. HttpTracing includes everything, including the normal things in the Tracing component (like the tracer and propagation), and the replacement for SpanNameProvider.

HttpTracing is nice because it should work nicely in any dependency injection system, like Spring, Guice or CDI, enclosing the non-instrumentation specific tracing configuration.

I haven't doc'd this class yet, but I would just note that the client and server parsers are where you do things like add tags based on servlet principal or http headers

**Comment by adriancole** [comment]
added tests to ensure grpc is compatible with both v1.2 and v1.3

**Comment by adriancole** [comment]
rewrote all the readmes.

waiting on this to merge https://github.com/openzipkin/zipkin/pull/1580

need to add http redirect tests, as I'm pretty sure some clients will have unexpected behavior

File: instrumentation/servlet/src/main/java/brave/servlet/TracingFilter.java (hunk 12 above)
Comment (adriancole) [inline instrumentation/servlet/src/main/java/brave/servlet/TracingFilter.java]: ok thrash is mostly over. Only thing I'm thinking of tuning is how remote addresses are parsed.

**Comment by adriancole** [comment]
renaming httpclient4 to httpclient since apache has made a new artifact httpclient5 (although not released anything on that in over a year)

**Comment by adriancole** [comment]
ok httpasync client works.

TODO:
 * Add tests for http client redirects
 * Revisit exception parsing
 * Revisit remote address parsing

File: instrumentation/jaxrs2/src/test/java/brave/jaxrs2/ITTracingFeature_Client.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  @Override
+  @Test(expected = AssertionError.class) // doesn't yet close a span on exception
+  public void addsErrorTagOnTransportException() throws Exception {
+    super.addsErrorTagOnTransportException();
+  }
+
+  @ClientAsyncExecutor
+  static class TracingExecutorServiceProvider implements ExecutorServiceProvider {
```
Comment (adriancole) [inline instrumentation/jaxrs2/src/test/java/brave/jaxrs2/ITTracingFeature_Client.java]: cc @jplock this works now..

**Comment by adriancole** [comment]
redirect test indeed broke a couple of the clients... fixing now

File: instrumentation/httpasyncclient/src/main/java/brave/httpasyncclient/TracingHttpAsyncClientBuilder.java [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+   *
+   * @see Constants#SERVER_ADDR
+   */
+  public TracingHttpAsyncClientBuilder remoteServiceName(String remoteServiceName) {
+    if (remoteServiceName == null) throw new NullPointerException("remoteServiceName == null");
+    this.remoteServiceName = remoteServiceName;
+    return this;
+  }
+
+  @Override public CloseableHttpAsyncClient build() {
```
Comment (adriancole) [inline instrumentation/httpasyncclient/src/main/java/brave/httpasyncclient/TracingHttpAsyncClientBuilder.java]: @shakuzen heh no comments yet.. but none needed, right? :p

**Comment by adriancole** [comment]
ok epic work to get redirects operating with expected output (span per attempt) is in.

next up is polishing exception handling a remote address parsing. final lap

**Comment by adriancole** [comment]
ok I have hidden everything not required to expose. tests are good, and I think the code is, too.

I have to document things tonight, but codewise, would appreciate any valiance in review, notably the customization model for HttpTracing. Here's a unit test of that.

For example, to override the naming policy for spans, you can do this, which will affect any http client:
```java
    httpTracing = httpTracing.toBuilder()
        .clientParser(new HttpClientParser() {
          // note you can take advantage of the actual request. ex req instanceof HttpServletRequest!
          @Override public <Req> String spanName(HttpAdapter<Req, ?> adapter, Req req) {
            return adapter.method(req).toLowerCase() + " " + adapter.path(req);
          }
        })
        .build();
```

Then when you are configuring a specific http client, you can assign the server name like so:

```java
    httpTracing = httpTracing.toBuilder().serverName("remote-service").build();
```

**Comment by adriancole** [comment]
added a few client benchmarks.. I'll add the others including server-side later.

**Comment by adriancole** [comment]
fyi here's an example run of the few client benchmarks (they are real requests, but against a stupid server)

```java
Benchmark                                                                       Mode  Cnt   Score   Error  Units
b.httpasyncclient.ApacheHttpAsyncClientBenchmarks.client_get                    avgt   15  61.399 ± 2.494  us/op
b.httpasyncclient.ApacheHttpAsyncClientBenchmarks.tracedButUnsampledClient_get  avgt   15  67.709 ± 4.623  us/op
b.httpasyncclient.ApacheHttpAsyncClientBenchmarks.tracedClient_get              avgt   15  69.701 ± 1.161  us/op
b.httpclient.ApacheHttpClientBenchmarks.client_get                              avgt   15  38.741 ± 1.196  us/op
b.httpclient.ApacheHttpClientBenchmarks.tracedButUnsampledClient_get            avgt   15  40.383 ± 0.768  us/op
b.httpclient.ApacheHttpClientBenchmarks.tracedClient_get                        avgt   15  45.378 ± 0.615  us/op
b.okhttp3.OkHttpClientBenchmarks.client_get                                     avgt   15  48.136 ± 1.143  us/op
b.okhttp3.OkHttpClientBenchmarks.tracedButUnsampledClient_get                   avgt   15  52.169 ± 2.923  us/op
b.okhttp3.OkHttpClientBenchmarks.tracedClient_get                               avgt   15  54.037 ± 1.181  us/op
```

**Comment by adriancole** [comment]
ok all http client impls are now benchmarked. going to the http server ones next

**Comment by adriancole** [comment]
just pushed benchmarks for all the http implementations client and server alike!

**Comment by adriancole** [comment]
there's something up with the apachehc benchmark, as it crashes on socket problems immediately. All the others ran. Here's an example run

```
Benchmark                                                              Mode  Cnt    Score    Error  Units
b.httpasyncclient.ApacheHttpAsyncClientBenchmarks.client_get           avgt   15   72.363 ±  9.468  us/op
b.httpasyncclient.ApacheHttpAsyncClientBenchmarks.tracedClient_get     avgt   15   76.197 ±  2.698  us/op
b.httpasyncclient.ApacheHttpAsyncClientBenchmarks.unsampledClient_get  avgt   15   67.840 ±  0.797  us/op
b.jaxrs2.JaxRs2ClientBenchmarks.client_get                             avgt   15   90.393 ± 15.926  us/op
b.jaxrs2.JaxRs2ClientBenchmarks.tracedClient_get                       avgt   15  106.585 ± 18.322  us/op
b.jaxrs2.JaxRs2ClientBenchmarks.unsampledClient_get                    avgt   15   99.672 ± 13.117  us/op
b.jaxrs2.JaxRs2ServerBenchmarks.server_get                             avgt   15   87.435 ±  2.102  us/op
b.jaxrs2.JaxRs2ServerBenchmarks.tracedServer_get                       avgt   15   90.488 ±  1.378  us/op
b.jaxrs2.JaxRs2ServerBenchmarks.unsampledServer_get                    avgt   15   87.958 ±  1.392  us/op
b.okhttp3.OkHttpClientBenchmarks.client_get                            avgt   15   50.602 ±  0.936  us/op
b.okhttp3.OkHttpClientBenchmarks.tracedClient_get                      avgt   15   60.113 ±  2.676  us/op
b.okhttp3.OkHttpClientBenchmarks.unsampledClient_get                   avgt   15   55.833 ±  1.254  us/op
b.servlet.ServletBenchmarks.server_get                                 avgt   15   65.671 ±  0.932  us/op
b.servlet.ServletBenchmarks.tracedServer_get                           avgt   15   72.415 ±  1.698  us/op
b.servlet.ServletBenchmarks.unsampledServer_get                        avgt   15   66.801 ±  2.277  us/op
b.sparkjava.SparkBenchmarks.server_get                                 avgt   15   75.719 ±  1.139  us/op
b.sparkjava.SparkBenchmarks.tracedServer_get                           avgt   15   81.540 ±  1.892  us/op
b.sparkjava.SparkBenchmarks.unsampledServer_get                        avgt   15   77.176 ±  1.425  us/op
b.spring.web.RestTemplateBenchmarks.client_get                         avgt   15   67.928 ±  1.042  us/op
b.spring.web.RestTemplateBenchmarks.tracedClient_get                   avgt   15   77.949 ±  0.984  us/op
b.spring.web.RestTemplateBenchmarks.unsampledClient_get                avgt   15   74.508 ±  0.779  us/op
b.spring.webmvc.WebMvcBenchmarks.server_get                            avgt   15   97.270 ±  1.697  us/op
b.spring.webmvc.WebMvcBenchmarks.tracedServer_get                      avgt   15  100.218 ±  1.742  us/op
b.spring.webmvc.WebMvcBenchmarks.unsampledServer_get                   avgt   15  100.051 ±  3.035  us/op
```

**Comment by adriancole** [comment]
the jaxrs client benchmarks look suspiciously bad. I'll check and see if @jplock's advice about the annotation is at fault, or something else.

**Comment by adriancole** [comment]
actually even the untraced jaxrs-client perf is significantly worse than others. Maybe there's just some overhead that's implicit with the configuration involved.

File: instrumentation/p6spy/src/main/java/brave/p6spy/TracingJdbcEventListener.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+  final String remoteServiceName;
+
+  TracingJdbcEventListener(String remoteServiceName) {
+    this.remoteServiceName = remoteServiceName;
+  }
+
+  @Override public void onBeforeAnyExecute(StatementInformation info) {
+    Tracer tracer = Tracing.currentTracer();
+    if (tracer == null) return;
```
Comment (adriancole) [inline instrumentation/p6spy/src/main/java/brave/p6spy/TracingJdbcEventListener.java]: This method is special-cased variant of `Tracing.tracer() != null` Anyway the rationale was expanded on here https://github.com/openzipkin/brave/pull/379#discussion_r111049601

Comment (adriancole) [inline instrumentation/p6spy/src/main/java/brave/p6spy/TracingJdbcEventListener.java]: main point is some libraries optimize for routine use of static wiring, brave doesn't. In doing so things are a lot simpler (and it discourages people from using edge-case config like currentTracer routinely)

File: instrumentation/servlet/README.md [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+        .getServletContext()
+        .addFilter("TracingFilter", TracingFilter.create(tracing))
+        .addMappingForUrlPatterns(EnumSet.allOf(DispatcherType.class), true, "/*");
+  }
+
+  @Override
+  public void contextDestroyed(ServletContextEvent servletContextEvent) {
+  }
+}
+```
```
Comment (adriancole) [inline instrumentation/servlet/README.md]: thx will do

File: instrumentation/jaxrs2/src/main/java/brave/jaxrs2/TracingContainerFilter.java (hunk 9 above)
Comment (adriancole) [inline instrumentation/jaxrs2/src/main/java/brave/jaxrs2/TracingContainerFilter.java]: added javadoc on why, and we're free to figure out a different way!

```java
  /**
   * This implementation peeks to see if the request is async or not, which means {@link
   * PreMatching} cannot be used: pre-matching doesn't inject the resource info!
   */
```

File: instrumentation/jaxrs2/README.md [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+The `TracingClientFilter` adds trace headers to outgoing requests.
+`TracingContainerFilter` extracts trace state from incoming requests.
+Both report to Zipkin how long each request takes, along with relevant
+tags like the http url.
+
+## Configuration
+
+### Dependency Injection
+Tracing always needs an instance of type `brave.http.HttpTracing`
+configured. Make sure this are in place before proceeding.
```
Comment (adriancole) [inline instrumentation/jaxrs2/README.md]: done

File: instrumentation/servlet/README.md (hunk 17 above)
Comment (adriancole) [inline instrumentation/servlet/README.md]: done

**Comment by adriancole** [comment]
merging so that @jplock @llinder @devinsba and others can try it out for a few days via snapshots prior to release. As a reminder, the snapshot repository is http://oss.jfrog.org/artifactory/oss-snapshot-local

**Comment by adriancole** [comment]
ps rewrote the PR description so it is more accessible

**Comment by adriancole** [comment]
during integration testing with an example app, I noticed a glitch in how we handle scope. We should remove the span from scope *before* reporting the span, as otherwise it is confusing in logging case.

Heads up for a revision which tightens up the tests to make sure context is cleaned *before* reporting the span.

**Comment by adriancole** [comment]
https://github.com/openzipkin/brave/pull/393
